contracts/erc721/ERC721PreApproval.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2021 the ethier authors (github.com/divergencetech/ethier)
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../thirdparty/opensea/OpenSeaGasFreeListing.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+/// @notice Pre-approval of OpenSea proxies for gas-less listing
+/// @dev This wrapper allows users to revoke the pre-approval of their
+/// associated proxy and emits the corresponding events. This is necessary for
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
+    /// @dev It is important that Active remains at first position, since this
+    /// is the scenario that we are trying to optimize for.
+    enum State {
+        Active,
+        Inactive
+    }
+
+    /// @notice The state of the pre-approval for a given address
+    mapping(address => State) private state;
+
+    /// @dev Returns true if either standard `isApprovedForAll()` or if the
+    /// `operator` is the OpenSea proxy for the `owner` provided the
+    /// pre-approval is active.
+    function isApprovedForAll(address owner, address operator)
+        public
+        view
+        virtual
+        override
+        returns (bool)
+    {
+        if (super.isApprovedForAll(owner, operator)) {
+            return true;
+        }
+
+        if (state[owner] == State.Active) {
+            return OpenSeaGasFreeListing.isApprovedForAll(owner, operator);
+        }
+
+        return false;
+    }
+
+    /// @dev Uses the standard `setApprovalForAll` or toggles the pre-approval
+    /// state if `operator` is the OpenSea proxy for the sender.
+    function setApprovalForAll(address operator, bool approved)
+        public
+        virtual
+        override
+    {
+        address owner = _msgSender();
+        if (operator == OpenSeaGasFreeListing.proxyFor(owner)) {
+            state[owner] = approved ? State.Active : State.Inactive;
+            emit ApprovalForAll(owner, operator, approved);
+        } else {
+            super._setApprovalForAll(owner, operator, approved);
+        }
+    }
+
+    /// @dev Checks if the receiver has an existing proxy. If not, the
+    /// pre-approval is disabled.
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override {
+        super._beforeTokenTransfer(from, to, tokenId);
+
+        // Exclude burns
+        if (to != address(0) && state[to] == State.Active) {
```

Comment:
```
if (to == address(0) || state[to] != State.Active) {
  return
}
```

Rationale: https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html TL;DR it acts as a signal to the reader of the code that nothing more will happen in this branch.